### DISCUSSION TITLE
Question on ConvexHull use

### DISCUSSION THREAD

**Discussion by jkissi** [marcomusy/vedo#642] [Q&A] [answered]
Hi Marco!
I am having a perplexing problem that I am totally stumped on. I am attempting to take a small number of points of points and draw a line connecting the points along the outer most edge only. I believe the best tool to use for this in vedo is the ConvexHull class. It seems very simple to use, I looked at the docs and the example works for me. 

When I attempt to pass it a numpy array of 3D points, it doesn't draw anything on the renderer and when I query the `ConvexHull() ` object, it says that it has no points associated with it:

```
ipdb> test_points
array([[-1.45738758, -1.81266949,  0.08192889],
       [-1.38937994,  0.885967  ,  0.08192889],
       [-1.38861339,  0.88480673,  0.08192889],
       [ 1.23244707, -1.9349347 ,  0.08192889],
       [ 1.32785031,  0.82507669,  0.08192889]]) 

ipdb> spl = ConvexHull(test_points)

ipdb> spl.points()
array([], dtype=float64)
```

Any idea why I could be getting this behaviour? Do I have to do something to the points beforehand?
Cheers
j

**Comment by marcomusy** [comment]
Hi Mr. J :)
that is strange, can you try this
```python
from vedo import *

test_points = np.array([
       [-1.45738758, -1.81266949,  0.08192889],
       [-1.38937994,  0.885967  ,  0.08192889],
       [-1.38861339,  0.88480673,  0.08192889],
       [ 1.23244707, -1.9349347 ,  0.08192889],
       [ 1.32785031,  0.82507669,  0.08192889],
]) 

spl = ConvexHull(test_points).lw(1)
show(spl, axes=1)
```
<img width="394" alt="Screenshot 2022-05-08 at 20 59 00" src="https://user-images.githubusercontent.com/32848391/167311458-d0a7b3c0-1968-4ae4-acd4-e2fdd551b18c.png">

**Reply by jkissi** [reply]
howdy Mr M!! ok this works fine, just like the example did. So i need to give you more context.
This problem is actually related to my previous question #640 , in that it is part for the function call `self.generateSpline()`. The idea is to replace all the individual lines that make up the straight line shape with a single spline, where the vertices are the numpy array of points I was talking about:

```
def generateSpline(self):
            """Helper function to generate closed spline from a series of connected lines
            """

            # make sure only the most important points are retained
            if len(self.cpoints) > len(self.new_arr):
                self.sfidx_curr = len(self.cpoints)

                if self.sfidx_curr > 0:
                    self.new_arr.append(self.cpoints[self.sfidx_curr-1])
                self.cpoints = []
                # deep copy the list, dont simply create a new pointer to the old list
                self.cpoints = self.new_arr.copy()

            # if the first and last points in cpoints are close together, create a closed spline
            if self.cpoints and mag(self.cpoints[0] - self.cpoints[-1]) < self.mesh.diagonalSize()*self.tol:
                # get the unique values
                unique, counts = np.unique(self.cpoints, axis=0, return_counts=True)
                self.cpoints = unique
                # THIS IS THE OFFENDING LINE
                spline = ConvexHull(self.cpoints).lw(1)
                print(f"This is the final number of cpoints: {len(self.cpoints)}")

                return spline
            else:
                return
```
If I put a breakpoint on `return spline`, then query self.cpoints (my list of vertices), spline (the ConvexHull() object) and spline.points(), I get the following:

```
ipdb> self.cpoints
array([[-0.67839045,  0.75197982,  0.43835315],
       [-0.65972665,  0.73331603,  0.43835315],
       [-0.64106286, -1.20771862,  0.43835315],
       [ 1.2439804 ,  0.77064362,  0.43835315],
       [ 1.46172468, -1.23882494,  0.43835315]])

ipdb> spline
(Mesh)00000274A1891CA0

ipdb> spline.points()
array([], dtype=float64)
```
If I then manually add the spline to the renderer and then query self.actors:
```
ipdb> self.add(spline)
<__main__.testo3d.<locals>.squareCutPlotter object at 0x00000274945B3BE0>

ipdb> self.actors
[(Text2D)000002749E8B9880, (Points)000002749E88B640, (Box)000002749E87D040, (Mesh)00000274A1891CA0]
```
Its adds the spline but the renderer shows no lines at all. 

This is a much more detailed account of what is happening. I thought that perhaps there was some nuance that you would know about but obviously, the strange behaviour is with my particular case, so more detail!

I am very confused as to what is happening here. hopefully you might be able to see something that I don't...

**Reply by marcomusy** [reply]
maybe it's a scope problem?
What if you add
```
                print(f"This is the final number of cpoints: {len(self.cpoints)}")
                print(f"This is the final number of splinepoints: {spline.points()}")
```

**Reply by jkissi** [reply]
I get the following output:
```
This is the final number of cpoints: 5
This is the final number of splinepoints: []
```

Did you mean that I should put the print statements next to eachother? Or was I supposed to place the splinepoints statement somewhere else?

**Reply by marcomusy** [reply]
next to each other, one right after the other in the same call function.

Besides, why do you call spline a ConvexHull? That's a polygonal closed surface, not a line. The fact that all z are the same can actually cause problems.

**Reply by jkissi** [reply]
....and there's the missing piece of information. If all the z coords are the same, it may cause problems? OK, I thought that because it works with 2D as well as 3D, that it would be able to handle that, but perhaps it can't and whatever that problem is, it ends up producing no output. 

OK, so lets assume that `ConvexHull()` won't solve my problem. How would I go about drawing a line around the outer most edge of a series of points here? Does vedo have a function that will do that for me reliably or do I have to create one myself? The `Line()` function tended to join the points by the shortest distance (I think), so the final line could end up all over the place. That's why I thought of `ConvexHull()`  to start with...

**Comment by marcomusy** [comment, marked as answer]
you should use `delaunay2D()`!

> The Line() function tended to join the points by the shortest distance (I think),

No! `Line` generates the line in the same order you give the points it joins.

**Reply by jkissi** [reply]
Hi Marco, 
So...after meditating on your last statement, I went back to the `Line()` function (I did look at `delaunay2D()` but it seemed to want to places lines on the inside of the shape and when I fed it the list of point indexes using the `boundaries` keyword, it crashed my kernel a couple of times, so I took that as a sign...;-)). From what you said, I realised that the function I was using to get unique point values for the final spline was also _reordering_ the points, resulting in the squiggly mess I was getting. Once I preserved the original order, everything finally worked as desired!

self.actors has one `Line()`
```
ipdb> self.actors
[(Text2D)000001BD890CDEE0, (Points)000001BD8B39C220, (Box)000001BD8B3BCD60, (Line)000001BD8B3ACF40]
```
and that `Line()` has the necessary points:

```
ipdb> self.actors[3].points()
array([[-1.61845949,  0.4132963 ,  0.13607348],
       [-0.01839863,  0.38379523,  0.16107396],
       [ 1.07289656, -0.1295091 ,  0.20775137],
       [ 0.29042357, -0.81126018,  0.23734586],
       [-1.67310939, -0.8371636 ,  0.21039703],
       [-2.14360832, -0.07914366,  0.15803139],
       [-1.96527958,  0.58989846,  0.1204299 ],
       [-1.60825483,  0.41523332,  0.13610526]])
```
It works with the original functionality to highlight the points when selected and crucially when I press 'c', it deletes the spline, meaning its usable. Blimey, that was a bit of an adventure. 

Anyway, thanks for you help and patience with this one (especially answering my questions on a Sunday, I genuinely was not expecting that...)
I made a little video, so you can see what I've been talking about. It's still janky, but it's the foundation I need to build on. 

Thanks again, you're an absolute legend. If you have any suggestions, they are always welcome. 
j

https://user-images.githubusercontent.com/8246894/167518026-328ccc3d-ed26-4668-87ba-546085ba0ab3.mp4

**Reply by marcomusy** [reply]
Sounds great, well done :)
one small thing i noticed in the video, you dont need to try to "close" the line by clicking near the first point, in principle the thin red line is already cutting the object when pressing z
moreover if you click past the first point, thus creating a "figure of eight" e.i. a crossing of the lines, the cutter might misbehave..

Thanks for the feedback!

**Reply by jkissi** [reply]
Hmmm, good points. so do you think I should adopt a different strategy with it? Currently I use the tolerance measure in the original code to see if the first and last points are close enough. If they are, I create a single line of all the points, do you think  should remove this?

**Reply by marcomusy** [reply]
well I guess that you dont need that, you can create a closed line (as you already do).
you may want to check if the line crosses itself somewhere.. but I would leave that as i refinement for later.. also it would be fair enough to just warn a user that s/he should not do it..:)

**Reply by jkissi** [reply]
OK then. I'll make sure to account for these when I flesh out the class properly. You'll probably end up hearing about it ;-). Thanks for the tips!!
